### PR TITLE
Fix RmsNorm backward pass for contiguous tensors (issue #2168)

### PR DESCRIPTION
**Pull request by Gravirus**
﻿## Fix RmsNorm backward pass for contiguous tensors (Issue #2168)

### Problem

In candle-nn 0.10.2, RmsNorm::forward has a fast path for contiguous tensors that uses apply_op2_no_bwd, which does not register a backward pass in the autograd graph. Since most tensors in a typical model are contiguous in memory, this means gradients silently stop flowing through RmsNorm during training.

The bug affects all models using RmsNorm from candle-nn, including LLaMA, Mistral, Qwen, Phi, Gemma, and others.

### Root Cause

Two interconnected issues:

1. layer_norm.rs: RmsNorm::forward checks xs.is_contiguous() and routes to ops::rms_norm() for contiguous tensors
2. ops.rs: ops::rms_norm() uses apply_op2_no_bwd and the RmsNorm CustomOp2 has no bwd method

### Fix

Two changes in two files:

**candle-nn/src/layer_norm.rs**: Removed the is_contiguous() fast path. RmsNorm::forward now always delegates to LayerNorm::forward (with remove_mean=false), which uses elementary operations (sqr, sum_keepdim, broadcast_div, broadcast_mul) that automatically build the autograd graph.

**candle-nn/src/ops.rs**: Added a bwd method to the RmsNorm CustomOp2 implementation with proper gradient formulas, and changed apply_op2_no_bwd to apply_op2 so the fast path can be used with gradient tracking in the future.

### Changes

`diff
// layer_norm.rs
impl Module for RmsNorm {
    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
-       if xs.is_contiguous() {
-           crate::ops::rms_norm(xs, &self.0.weight, self.0.eps as f32)
-       } else {
-           self.0.forward(xs)
-       }
+       self.0.forward(xs)
    }
}
`

`diff
// ops.rs
impl candle::CustomOp2 for RmsNorm {
    // ... existing cpu_fwd, cuda_fwd, metal_fwd ...
    
+   fn bwd(
+       &self,
+       arg1: &Tensor,
+       arg2: &Tensor,
+       _res: &Tensor,
+       grad_res: &Tensor,
+   ) -> Result<(Option<Tensor>, Option<Tensor>)> {
+       let n = arg1.dim(D::Minus1)? as f32;
+       let rms = (arg1.sqr()?.sum_keepdim(D::Minus1)? / n + self.eps)?.sqrt()?;
+       let alpha_over_rms = arg2.broadcast_div(&rms)?;
+       let x_dot_grad_alpha = (arg1 * grad_res * arg2)?.sum_keepdim(D::Minus1)?;
+       let correction = arg1 * &x_dot_grad_alpha / (rms.sqr() * n)?;
+       let grad_x = alpha_over_rms.broadcast_mul(&(grad_res - correction)?)?;
+       let grad_alpha = (arg1 * grad_res / rms)?.sum_keepdim(D::Minus1)?;
+       Ok((Some(grad_x), Some(grad_alpha)))
+   }
 }

 pub fn rms_norm(xs: &Tensor, alpha: &Tensor, eps: f32) -> Result<Tensor> {
     // ... validation ...
-    xs.apply_op2_no_bwd(alpha, &RmsNorm { eps })
+    xs.apply_op2(alpha, RmsNorm { eps })
 }
`

### Testing

Tested on a Qwen2.5-0.5B model with DynamicLoRA adapters:

| Metric | Before Fix | After Fix |
|--------|-----------|-----------|
| Variables with gradients | 0/6 | 6/6 |
| Training loss (epoch 0) | 21.23 | 21.23 |
| Training loss (epoch 39) | N/A (no learning) | 2.78 |
| Generation quality | Random tokens | Coherent output |

### Related Issues

- Issue #2168: No backward pass for RmsNorm if tensor is contiguous (OPEN since May 2024)
- Issue #2977: No backward pass for LayerNorm if tensor is contiguous
- Issue #3011: LayerNorm Gradient Flow Issue in candle-nn

### CODE REVIEW

**Comment by nxrobins** [comment]
Independent corroboration: this bug also surfaced during Phase 0.5b validation of an external project against a Python reference implementation, on the same model you tested (Qwen2.5-0.5B). Confirming both the symptom and the fix.

**Reproducer (standalone, no external assets, <60s):** https://github.com/nxrobins/candle-bug-3-rmsnorm-no-bwd

\`\`\`bash
git clone https://github.com/nxrobins/candle-bug-3-rmsnorm-no-bwd && cd candle-bug-3-rmsnorm-no-bwd && cargo run --release
# exit 1 at candle main HEAD 1d7e9274: BUG PRESENT
# exit 0: bug fixed
\`\`\`

The reproducer pins to commit \`7c7a8c570e\` (the 0.10.2 release commit), builds a tiny \`RmsNorm\` via the standard \`candle_nn::rms_norm\` constructor, passes a contiguous \`Var\` through \`RmsNorm::forward\`, and asserts that \`grads.get(var)\` returns \`None\`. Positive control: \`forward_diff\` propagates gradient (\`max(|g|) = 0.24\` on this fixture).

This pattern is also present in two sibling functions that PR #3526 doesn't address:
- \`candle-nn/src/rotary_emb.rs:537\` (\`rope\` and friends) — filed as #3568
- \`candle-nn/src/ops.rs:430\` (\`softmax_last_dim\`) — filed as #3569

Both filed independently with the same \`bwd\`-on-CustomOp shape as this PR as the suggested supremum fix; they reference this PR as precedent for the approach. If the maintainers prefer one bundled PR addressing all four (RmsNorm + rope + softmax + an open one for Qwen2 causal mask, #3567), happy to help coordinate.

Full validation context with all four bugs that surfaced together: https://github.com/nxrobins/wave-agent/blob/main/docs/upstream-candle-bugs.md
